File: shared/src/main/java/io/github/zwieback/familyfinance/util/DateUtils.java
package io.github.zwieback.familyfinance.util;

import android.content.Intent;
import android.os.Bundle;
import android.os.Parcel;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.threeten.bp.LocalDate;
import org.threeten.bp.format.DateTimeParseException;
import org.threeten.bp.temporal.TemporalAdjusters;

import java.util.Calendar;

import static io.github.zwieback.familyfinance.util.NumberUtils.ID_AS_NULL;
import static io.github.zwieback.familyfinance.util.NumberUtils.isNullId;
import static io.github.zwieback.familyfinance.util.StringUtils.EMPTY;
import static io.github.zwieback.familyfinance.util.StringUtils.isTextEmpty;

public final class DateUtils {

    // bundle keys
    private static final String KEY_YEAR = "keyYear";
    private static final String KEY_MONTH = "keyMonth";
    private static final String KEY_DAY_OF_MONTH = "keyDayOfMonth";

    // workaround for month of calendar (start with 0)
    private static final int MONTH_OF_CALENDAR_INCREMENT = 1;

    public static boolean isTextAnLocalDate(@Nullable String text) {
        try {
            return stringToLocalDate(text) != null;
        } catch (DateTimeParseException e) {
            return false;
        }
    }

    @Nullable
    public static LocalDate stringToLocalDate(@Nullable String text) {
        if (isTextEmpty(text)) {
            return null;
        }
        return LocalDate.parse(text);
    }

    @NonNull
    public static String localDateToString(@Nullable LocalDate localDate) {
        if (localDate == null) {
            return EMPTY;
        }
        return localDate.toString();
    }

    @NonNull
    public static Calendar localDateToCalendar(@Nullable LocalDate localDate) {
        if (localDate == null) {
            return Calendar.getInstance();
        }
        Calendar calendar = Calendar.getInstance();
        calendar.set(localDate.getYear(), localDate.getMonthValue() - MONTH_OF_CALENDAR_INCREMENT,
                localDate.getDayOfMonth());
        return calendar;
    }

    @NonNull
    public static LocalDate calendarDateToLocalDate(int calendarYear,
                                                    int calendarMonth,
                                                    int calendarDay) {
        return LocalDate.of(calendarYear, calendarMonth + MONTH_OF_CALENDAR_INCREMENT, calendarDay);
    }

    @NonNull
    public static LocalDate now() {
        return LocalDate.now();
    }

    @NonNull
    public static LocalDate startOfMonth() {
        return now().with(TemporalAdjusters.firstDayOfMonth());
    }

    @NonNull
    public static LocalDate endOfMonth() {
        return now().with(TemporalAdjusters.lastDayOfMonth());
    }

    public static void writeLocalDateToBundle(Bundle out, @NonNull LocalDate date) {
        out.putInt(KEY_YEAR, date.getYear());
        out.putInt(KEY_MONTH, date.getMonthValue());
        out.putInt(KEY_DAY_OF_MONTH, date.getDayOfMonth());
    }

    @NonNull
    public static LocalDate readLocalDateFromBundle(Bundle in) {
        int year = in.getInt(KEY_YEAR);
        int month = in.getInt(KEY_MONTH);
        int dayOfMonth = in.getInt(KEY_DAY_OF_MONTH);
        return LocalDate.of(year, month, dayOfMonth);
    }

    public static void writeLocalDateToIntent(@NonNull Intent out,
                                              @NonNull String name,
                                              @NonNull LocalDate date) {
        out.putExtra(name + KEY_YEAR, date.getYear());
        out.putExtra(name + KEY_MONTH, date.getMonthValue());
        out.putExtra(name + KEY_DAY_OF_MONTH, date.getDayOfMonth());
    }

    @NonNull
    public static LocalDate readLocalDateFromIntent(@NonNull Intent in, @NonNull String name) {
        LocalDate today = now();
        int year = in.getIntExtra(name + KEY_YEAR, today.getYear());
        int month = in.getIntExtra(name + KEY_MONTH, today.getMonthValue());
        int dayOfMonth = in.getIntExtra(name + KEY_DAY_OF_MONTH, today.getDayOfMonth());
        return LocalDate.of(year, month, dayOfMonth);
    }

    public static void writeLocalDateToParcel(Parcel out, @Nullable LocalDate date) {
        if (date == null) {
            out.writeInt(ID_AS_NULL);
            out.writeInt(ID_AS_NULL);
            out.writeInt(ID_AS_NULL);
        } else {
            out.writeInt(date.getYear());
            out.writeInt(date.getMonthValue());
            out.writeInt(date.getDayOfMonth());
        }
    }

    @Nullable
    public static LocalDate readLocalDateFromParcel(Parcel in) {
        int year = in.readInt();
        int month = in.readInt();
        int dayOfMonth = in.readInt();
        if (isNullId(year) || isNullId(month) || isNullId(dayOfMonth)) {
            return null;
        }
        return LocalDate.of(year, month, dayOfMonth);
    }

    private DateUtils() {
    }
}

File: shared/src/main/java/io/github/zwieback/familyfinance/util/DateUtils.kt
package io.github.zwieback.familyfinance.util

import io.github.zwieback.familyfinance.constant.DateConstants.BANK_DATE_FORMATTER
import io.github.zwieback.familyfinance.constant.DateConstants.EPOCH_DATE
import org.threeten.bp.LocalDate
import org.threeten.bp.format.DateTimeFormatter
import org.threeten.bp.format.DateTimeParseException
import org.threeten.bp.temporal.ChronoUnit
import org.threeten.bp.temporal.IsoFields
import org.threeten.bp.temporal.TemporalAdjusters
import java.util.*

object DateUtils {

    // workaround for month of calendar (start with 0)
    private const val MONTH_OF_CALENDAR_INCREMENT = 1


    @JvmStatic
    fun isTextAnLocalDate(text: String?): Boolean {
        return try {
            stringToLocalDate(text) != null
        } catch (e: DateTimeParseException) {
            false
        }
    }

    @JvmStatic
    fun stringToLocalDate(text: String?): LocalDate? {
        return if (!text.isNullOrEmpty()) {
            LocalDate.parse(text)
        } else {
            null
        }
    }

    @JvmStatic
    fun bankDateToLocalDate(text: String?): LocalDate {
        return if (!text.isNullOrEmpty()) {
            LocalDate.from(BANK_DATE_FORMATTER.parse(text))
        } else {
            now()
        }
    }

    @JvmStatic
    fun localDateToString(localDate: LocalDate?): String {
        return localDate?.toString().orEmpty()
    }

    @JvmStatic
    fun localDateToString(localDate: LocalDate?, formatter: DateTimeFormatter): String {
        return localDate?.format(formatter).orEmpty()
    }

    @JvmStatic
    fun localDateToCalendar(localDate: LocalDate?): Calendar {
        return localDate?.let {
            Calendar.getInstance().apply {
                set(
                    localDate.year,
                    localDate.monthValue - MONTH_OF_CALENDAR_INCREMENT,
                    localDate.dayOfMonth
                )
            }
        } ?: Calendar.getInstance()
    }

    @JvmStatic
    fun calendarDateToLocalDate(
        calendarYear: Int,
        calendarMonth: Int,
        calendarDay: Int
    ): LocalDate {
        return LocalDate.of(calendarYear, calendarMonth + MONTH_OF_CALENDAR_INCREMENT, calendarDay)
    }

    @JvmStatic
    fun now(): LocalDate {
        return LocalDate.now()
    }

    @JvmStatic
    fun startOfMonth(): LocalDate {
        return now().with(TemporalAdjusters.firstDayOfMonth())
    }

    @JvmStatic
    fun endOfMonth(): LocalDate {
        return now().with(TemporalAdjusters.lastDayOfMonth())
    }

    @JvmStatic
    fun localDateToEpochDay(date: LocalDate): Long {
        return ChronoUnit.DAYS.between(EPOCH_DATE, date)
    }

    @JvmStatic
    fun epochDayToLocalDate(daysFromEpoch: Long): LocalDate {
        return EPOCH_DATE.plusDays(daysFromEpoch)
    }

    @JvmStatic
    fun localDateToEpochWeek(date: LocalDate): Long {
        return ChronoUnit.WEEKS.between(EPOCH_DATE, date)
    }

    @JvmStatic
    fun epochWeekToLocalDate(monthsFromWeek: Long): LocalDate {
        return EPOCH_DATE.plusWeeks(monthsFromWeek)
    }

    @JvmStatic
    fun localDateToEpochMonth(date: LocalDate): Long {
        return ChronoUnit.MONTHS.between(EPOCH_DATE, date)
    }

    @JvmStatic
    fun epochMonthToLocalDate(monthsFromEpoch: Long): LocalDate {
        return EPOCH_DATE.plusMonths(monthsFromEpoch)
    }

    @JvmStatic
    fun localDateToEpochQuarter(date: LocalDate): Long {
        return IsoFields.QUARTER_YEARS.between(EPOCH_DATE, date)
    }

    @JvmStatic
    fun epochQuarterToLocalDate(quartersFromWeek: Long): LocalDate {
        return plusQuarters(EPOCH_DATE, quartersFromWeek)
    }

    @JvmStatic
    fun plusQuarters(date: LocalDate, numberOfQuarters: Long): LocalDate {
        return date.plus(numberOfQuarters, IsoFields.QUARTER_YEARS)
    }

    @JvmStatic
    fun extractWeekOfYear(date: LocalDate): Int {
        return date.get(IsoFields.WEEK_OF_WEEK_BASED_YEAR)
    }

    @JvmStatic
    fun extractQuarterOfYear(date: LocalDate): Int {
        return date.get(IsoFields.QUARTER_OF_YEAR)
    }
}
